tiledb/cloud/_results/results.py

```diff
@@ -68,14 +78,54 @@ def to_stored_param(self) -> stored_params.StoredParam:
             task_id=self.task_id,
         )
 
-    # The HTTP content of the body that was returned.
-    body: bytes
-    # The server-generated UUID of the task.
-    task_id: Optional[uuid.UUID]
-    # The decoder that was used to decode the results.
-    decoder: decoders.AbstractDecoder[_T]
-    # True if the results were stored, false otherwise.
-    results_stored: bool
+    task_id = attr.ib(type=Optional[uuid.UUID])
+    """The server-generated UUID of the task."""
+
+    decoder = attr.ib(type=decoders.AbstractDecoder[_T])
+    """The Decoder that was used to decode the results."""
+
+    results_stored = attr.ib(type=bool)
+    """True if the results were stored, false otherwise."""
+
+    _body = attr.ib(type=Optional[bytes])
+    """The HTTP content of the body that was returned.
+
+    If None, the result was not downloaded (yet).
+    """
+
+    @_body.validator
+    def _validate(self, attribute, value):
+        del attribute, value  # unused
+        if self.results_stored and not self.task_id:
+            raise ValueError("task_id must be set for stored results")
+        if not self.results_stored and not self._body:
```

Comment:
Is the validator called (1) only in the constructor and (2) after `_body` is actually set on the object?

Reply:
This is called only at initialization time.

Reply:
What I'm wondering is when `_body` is set here -- does attrs set it first _and then_ call the validator?

Reply:
Oh, yes. Validators are run only after the rest of the `__init__` method has completed, so at that point the class is fully set up, and validation ensures that the setup is correct before it is returned to the caller.

Reply:
Thanks